bindings/azure/openai/openai.go

```diff
@@ -0,0 +1,283 @@
+	p.client, err = azopenai.NewClientWithKeyCredential(m.APIBase, keyCredential, m.DeploymentId, nil)
+	if err != nil {
+		// TODO: handle error
+		return fmt.Errorf("error in connecting to openai: %w", err)
+	}
+
+	return nil
+}
+
+// Operations returns list of operations supported by OpenAI binding.
+func (p *AzOpenAI) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{
+		ChatCompletionOperation,
+		CompletionOperation,
+	}
+}
+
+// Invoke handles all invoke operations.
+func (p *AzOpenAI) Invoke(ctx context.Context, req *bindings.InvokeRequest) (resp *bindings.InvokeResponse, err error) {
+	p.logger.Infof("Invoke request: %v", req)
+	if req == nil {
+		return nil, errors.New("invoke request required")
+	}
+
+	if req.Metadata == nil {
+		return nil, errors.New("metadata required")
+	}
+	p.logger.Debugf("operation: %v", req.Operation)
+
+	startTime := time.Now().UTC()
+	resp = &bindings.InvokeResponse{
+		Metadata: map[string]string{
+			"operation":  string(req.Operation),
+			"start-time": startTime.Format(time.RFC3339Nano),
+		},
+	}
+
+	switch req.Operation { //nolint:exhaustive
+	case CompletionOperation:
+		response, err := p.completion(ctx, req.Data, req.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("error in the completion api %s: %w", &req.Data, err)
+		}
+		p.logger.Infof("response is : %v", response)
+		resp.Metadata["response"] = response
+
+	case ChatCompletionOperation:
+		response, err := p.chatCompletion(ctx, req.Data, req.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("error executing %s: %w", &req.Data, err)
+		}
+		p.logger.Infof("response is : %v", response)
+		resp.Metadata["response"] = response
+
+	default:
+		return nil, fmt.Errorf(
+			"invalid operation type: %s. Expected %s, %s, %s",
+			req.Operation, CompletionOperation, ChatCompletionOperation,
+		)
+	}
+
+	endTime := time.Now().UTC()
+	resp.Metadata["end-time"] = endTime.Format(time.RFC3339Nano)
+	resp.Metadata["duration"] = endTime.Sub(startTime).String()
+
+	return resp, nil
+}
+
+func (s *ChatSettings) Decode(in interface{}) error {
+	if err := config.Decode(in, s); err != nil {
+		return fmt.Errorf("decode failed. %w", err)
+	}
+
+	return nil
+}
+
+func (p *AzOpenAI) completion(ctx context.Context, message []byte, metadata map[string]string) (response string, err error) {
+	var ma Prompt
+	err = json.Unmarshal(message, &ma)
+	if err != nil {
+		return "", fmt.Errorf("error in unmarshalling the message array: %w", err)
+	}
+
+	settings := ChatSettings{}
+	settings.Temperature = 1.0
+	settings.TopP = 1.0
+	settings.MaxTokens = 16
+	settings.N = 1
+
+	err = settings.Decode(metadata)
+	if err != nil {
+		return "", fmt.Errorf("Error in decoding the parameters: %w", err)
+	}
+
+	resp, err := p.client.GetCompletions(context.TODO(), azopenai.CompletionsOptions{
+		Prompt:      []*string{to.Ptr(ma.Prompt)},
+		MaxTokens:   to.Ptr(int32(15)),
+		Temperature: to.Ptr(float32(0.0)),
```

Comment:
Shouldn't these come from the request?

Reply:
It should but these are optional properties with some default value set.